### PR TITLE
Further making Microsoft.AspNetCore.Sockets "not a thing"

### PR DESCRIPTION
**Pull request by davidfowl**
- This PR attempts to move things where they are needed instead of where they
happened to be used. As a result we should now have Sockets.Abstractions and
Sockets down to the minimal set of things required to make them run.
Sockets.Abstractions should go away in favor of Protocol.Abstractions and
Sockets contains the EndPoint abstraction and related types.
- Moved ConnectionManager and friends to
Sockets.Http.
-Removed Sockets and moved everything into Sockets.Abstractions.
- Moved DefaultConnection and put it in Sockets.Abstractions.

### CODE REVIEW

**Comment by davidfowl** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs [hunk 1]
```diff
@@ -22,7 +22,7 @@ public class ConnectionManager
         // TODO: Consider making this configurable? At least for testing?
         private static readonly TimeSpan _heartbeatTickRate = TimeSpan.FromSeconds(1);
 
-        private readonly ConcurrentDictionary<string, DefaultConnectionContext> _connections = new ConcurrentDictionary<string, DefaultConnectionContext>();
+        private readonly ConcurrentDictionary<string, (DefaultConnectionContext Connection, ValueStopwatch Timer)> _connections = new ConcurrentDictionary<string, (DefaultConnectionContext Connection, ValueStopwatch Timer)>();
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs]: 🎉

File: src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs [hunk 2]
```diff
@@ -53,7 +53,14 @@ public void Start()
 
         public bool TryGetConnection(string id, out DefaultConnectionContext connection)
         {
-            return _connections.TryGetValue(id, out connection);
+            connection = null;
+
+            if (_connections.TryGetValue(id, out var pair))
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs]: Would `return _connections.TryGetValue(id, out (connection, _));` work? Destructure all the things?!

File: src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs [hunk 3]
```diff
@@ -78,10 +84,10 @@ public DefaultConnectionContext CreateConnection()
 
         public void RemoveConnection(string id)
         {
-            if (_connections.TryRemove(id, out var connection))
+            if (_connections.TryRemove(id, out var pair))
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs]: As above, `if (_connections.TryRemove(id, out (_, timer)))`?

File: src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs [hunk 4]
```diff
@@ -128,33 +134,34 @@ public void Scan()
                 {
                     var status = DefaultConnectionContext.ConnectionStatus.Inactive;
                     var lastSeenUtc = DateTimeOffset.UtcNow;
+                    var connection = c.Value.Connection;
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs]: Destructure in the for loop! `foreach (var (connection, _) in _connections)`

File: src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs (hunk 3 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs]: why?

File: src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs (hunk 2 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs]: Ugh,

File: src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs (hunk 4 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs]: ....

File: src/Microsoft.AspNetCore.Sockets.Abstractions/ConnectionBuilderExtensions.cs [hunk 5]
```diff
@@ -1,16 +1,21 @@
...
-using Microsoft.Extensions.DependencyInjection;
 
 namespace Microsoft.AspNetCore.Sockets
 {
     public static class ConnectionBuilderExtensions
     {
         public static IConnectionBuilder UseEndPoint<TEndPoint>(this IConnectionBuilder connectionBuilder) where TEndPoint : EndPoint
         {
-            var endpoint = connectionBuilder.ApplicationServices.GetRequiredService<TEndPoint>();
+            var endpoint = (TEndPoint)connectionBuilder.ApplicationServices.GetService(typeof(TEndPoint));
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Sockets.Abstractions/ConnectionBuilderExtensions.cs]: Where is no generic `GetService`?

File: src/Microsoft.AspNetCore.Sockets.Abstractions/ConnectionBuilderExtensions.cs [hunk 6]
```diff
@@ -1,16 +1,21 @@
...
+
+            if (endpoint == null)
+            {
+                throw new InvalidOperationException($"{nameof(EndPoint)} type {typeof(TEndPoint)} is not registered.");
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Sockets.Abstractions/ConnectionBuilderExtensions.cs]: Is there a test that throws this error and checks for this error message?

File: src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs (hunk 3 above)
Comment (JamesNK) [reply src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs]: Destructure looks sweet but IMO not as easy to parse as a reader.

File: src/Microsoft.AspNetCore.Sockets.Abstractions/ConnectionBuilderExtensions.cs (hunk 5 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Abstractions/ConnectionBuilderExtensions.cs]: I didn't want to add a dependency on dependency injection 😄

File: src/Microsoft.AspNetCore.Sockets.Abstractions/ConnectionBuilderExtensions.cs (hunk 6 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Abstractions/ConnectionBuilderExtensions.cs]: No. But now we're debating removing end points altogether. I think I might just do that..

Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Abstractions/ConnectionBuilderExtensions.cs]: This is a bigger change than I'd want to make in this PR.

**Pull request merged**